packages/gguf/src/gguf.ts

```diff
@@ -0,0 +1,275 @@
+type MetadataBaseValue = string | number | bigint | boolean;
+type MetadataValue = MetadataBaseValue | MetadataBaseValue[] | MetadataValue[]; /// recursive as arrays can be nested.
+
+type Version = 1 | 2 | 3;
+const isVersion = (version: number): version is Version => version === 1 || version === 2 || version === 3;
+
+const ggufMagicNumber = new Uint8Array([0x47, 0x47, 0x55, 0x46]); /// "GGUF"
+
+export enum GGMLQuantizationType {
+	F32 = 0,
+	F16 = 1,
+	Q4_0 = 2,
+	Q4_1 = 3,
+	Q5_0 = 6,
+	Q5_1 = 7,
+	Q8_0 = 8,
+	Q8_1 = 9,
+	Q2_K = 10,
+	Q3_K = 11,
+	Q4_K = 12,
+	Q5_K = 13,
+	Q6_K = 14,
+	Q8_K = 15,
+	IQ2_XXS = 16,
+	IQ2_XS = 17,
+	IQ3_XXS = 18,
+	IQ1_S = 19,
+	IQ4_NL = 20,
+	IQ3_S = 21,
+	IQ2_S = 22,
+	IQ4_XS = 23,
+}
+
+enum GGUFValueType {
+	UINT8 = 0,
+	INT8 = 1,
+	UINT16 = 2,
+	INT16 = 3,
+	UINT32 = 4,
+	INT32 = 5,
+	FLOAT32 = 6,
+	BOOL = 7,
+	STRING = 8,
+	ARRAY = 9,
+	UINT64 = 10,
+	INT64 = 11,
+	FLOAT64 = 12,
+}
+function isGGUFValueType(n: number): n is GGUFValueType {
+	return typeof GGUFValueType[n] === "string";
+}
+
+const HTTP_CHUNK_SIZE = 2 * 10 ** 6; /// 2MB
+const HTTP_DATA_LEEWAY = 5 * 10 ** 5; /// 500kb
+
+/**
+ * Internal stateful instance to fetch ranges of HTTP data when needed
+ */
+class RangeView {
+	private chunk: number;
+	private buffer: ArrayBuffer;
+
+	readonly view: DataView;
+
+	constructor(public url: string) {
+		this.chunk = 0;
+		/// TODO(fix typing)
+		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+		// @ts-ignore
+		this.buffer = new ArrayBuffer(0, { maxByteLength: 50 * 10 ** 6 });
+		this.view = new DataView(this.buffer);
+	}
+	/**
+	 * Fetch a new chunk from the server
+	 */
+	async fetchChunk() {
+		const range = [this.chunk * HTTP_CHUNK_SIZE, (this.chunk + 1) * HTTP_CHUNK_SIZE - 1];
+		const buf = new Uint8Array(
+			await (
+				await fetch(this.url, {
+					headers: {
+						Range: `bytes=${range[0]}-${range[1]}`,
+					},
+				})
+			).arrayBuffer()
+		);
+		/// TODO(fix typing)
+		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+		// @ts-ignore
+		this.buffer.resize((this.chunk + 1) * HTTP_CHUNK_SIZE);
+		new Uint8Array(this.buffer).set(buf, this.chunk * HTTP_CHUNK_SIZE);
+		this.chunk += 1;
+	}
+	/**
+	 * Check whether we need to fetch a new chunk
+	 */
+	async check(offset: number) {
```

Comment:
```suggestion
	async fetchChunkIfNeeded(offset: number) {
```

nit. otherwise, `async check(offset: number)` sounds like a function that should return `bool`

Reply:
9366d4ae27a0ff6e289b7f63280ff56d0c80a715